SAML2/XML/Signature.hs

```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
+          where
+            keyinfo (KeyInfoKeyValue kv) = publicKeyValues kv
+            keyinfo (X509Data l) = foldMap keyx509d l
+              where
+                keyx509d (X509Certificate sc) = keyx509p $ X509.certPubKey $ X509.getCertificate sc
+                keyx509d _ = mempty
+                keyx509p (X509.PubKeyRSA r) = mempty{ publicKeyRSA = Just r }
+                keyx509p (X509.PubKeyDSA d) = mempty{ publicKeyDSA = Just d }
+                keyx509p _ = mempty
+            keyinfo _ = mempty
+
+        alg :: IdentifiedURI SignatureAlgorithm
+        alg = signatureMethodAlgorithm $ signedInfoSignatureMethod signedInfoTyped
+
+        dig :: Base64Binary
+        dig = signatureValue $ signatureSignatureValue signatureElem
+
+    case verifyBytes keys alg dig signedInfoElem of
+      Nothing    -> throwError . SignatureVerificationCryptoUnsupported $ show (keys, alg, dig, signedInfoElem)
+      Just False -> throwError . SignatureVerificationCryptoFailed      $ show (keys, alg, dig, signedInfoElem)
       Just True  -> pure ()
 
-  where
-  child n = HXT.runLA $ HXT.getChildren HXT.>>> isDSElem n HXT.>>> HXT.cleanupNamespaces HXT.collectPrefixUriPairs
-  keyinfo (KeyInfoKeyValue kv) = publicKeyValues kv
-  keyinfo (X509Data l) = foldMap keyx509d l
-  keyinfo _ = mempty
-  keyx509d (X509Certificate sc) = keyx509p $ X509.certPubKey $ X509.getCertificate sc
-  keyx509d _ = mempty
-  keyx509p (X509.PubKeyRSA r) = mempty{ publicKeyRSA = Just r }
-  keyx509p (X509.PubKeyDSA d) = mempty{ publicKeyDSA = Just d }
-  keyx509p _ = mempty
-  xpathsel t = "/*[local-name()='" ++ t ++ "' and namespace-uri()='" ++ namespaceURIString ns ++ "']"
-  xpathbase = "/*" ++ xpathsel "Signature" ++ xpathsel "SignedInfo" ++ "//"
-  xpath = xpathbase ++ ". | " ++ xpathbase ++ "@* | " ++ xpathbase ++ "namespace::*"
+
+-- | if name spaces that are declared in doc and used in the signed subtree, we need to copy
+-- the declarations into the part the subtree to make it self-contained.  the easiest way to
+-- impleemnt that is to use a xml:dsig canonicalization function and do another few rounds of
+-- rendering and parsing.
+--
+-- TODO: there may be a cleaner way to do this.  i know that xml-conduit isn't very interested
+-- in getting name spaces right, and HXT doesn't seem to be very successful at it either, but
+-- it may just be that i'm unaware of the regions of the HXT jungle that do this.
+fixNamespaces :: HasCallStack => HXT.XmlTree -> IO HXT.XmlTree
+fixNamespaces doc = do
+  can :: SBS
+    <- liftIO . capture' "fixNamespaces" $
+       canonicalize (CanonicalXMLExcl10 False) Nothing Nothing $ DOM.mkRoot [] [doc]
+  maybe (throwIO . ErrorCall $ "parse error on canonicalized xml") pure $
+    xmlToDoc (cs can)
+
+capture' :: String -> IO a -> IO a
+capture' actionName action = hCapture [stdout, stderr] action >>= \case
+  ("", !out) -> pure out
+  (noise, _) -> throwIO . ErrorCall $ actionName <> ": " <> noise
```

Comment:
If any noise on stdout indicates an error; it's probably helpful to have the output of stderr appended here too.

Reply:
i think that's what this does, no?

(thanks for the other comments, more changes coming up!)

Reply:
Ahh; I see! I misunderstood; It looked at a glance that the tuple matched up with `[stdout, stderr]`; but in hindsight that doest make sense 👍